networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/compression/SnappyCompressor.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2.compression;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import org.apache.tuweni.bytes.Bytes;
+import org.xerial.snappy.SnappyFramedInputStream;
+import org.xerial.snappy.SnappyFramedOutputStream;
+import tech.pegasys.artemis.util.iostreams.DelegatingInputStream;
+
+public class SnappyCompressor implements Compressor {
+  // The max uncompressed bytes that will be packed into a single frame
+  // See:
+  // https://github.com/google/snappy/blob/251d935d5096da77c4fef26ea41b019430da5572/framing_format.txt#L104-L106
+  static final int MAX_FRAME_CONTENT_SIZE = 65536;
+
+  @Override
+  public Bytes compress(final Bytes data) {
+
+    try (final ByteArrayOutputStream out = new ByteArrayOutputStream(data.size() / 2);
+        final OutputStream compressor = new SnappyFramedOutputStream(out)) {
+      compressor.write(data.toArrayUnsafe());
+      compressor.flush();
+      return Bytes.wrap(out.toByteArray());
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to compress data", e);
+    }
+  }
+
+  @Override
+  public Bytes uncompress(final InputStream input, final int maxBytes) throws CompressionException {
+    // This is a bit of a hack - but we don't want to close the underlying stream when
+    // we close the SnappyFramedInputStream
+    final UnclosableInputStream wrappedStream = new UnclosableInputStream(input);
+
+    try (final InputStream snappyIn = new SnappyFramedInputStream(wrappedStream)) {
+      return Bytes.wrap(snappyIn.readNBytes(maxBytes));
+    } catch (IOException e) {
```

Comment:
The potential risk here is that `SnappyFramedInputStream` maintains any kind of internal state or buffer of read data.  Sounds like it works now but it feels slightly dangerous to be creating and discarding wrapper input streams.

Ideally when the input stream was first created, the compressor would get a chance to wrap it and then you could just use it as an input stream to read data from.  It might be possible to do that in the `RpcRequestHandler` but not entirely sure.

Reply:
I can rework the javadoc to make it clear that the `maxBytes` param is really the expected size of the final payload to uncompress.  If the `SnappyFramedInputStream` ends up pulling more data than expected from the underlying stream, it means that there's extra data appended that we don't expect, so it's an error anyway.

The problem with wrapping the initial input stream is that we're not actually processing a series of snappy data.  For responses, we're processing a sequence of `statusCode|lengthPrefix|snappyData`.  So it's not really straightforward.  I guess we could wrap the entire inputStream, and then read from the underlying inputStream when we process the status and lengthPrefix, but that seems kind of messy too.

Reply:
I reworked the API, but it may be cleaner to try the other approach - I'll experiment.

Reply:
Hmm, yeah the fact that it's not all snappy data makes life much more interesting. I don't think it would be right to just always wrap it so maybe better to stick with what we have then.

The completely correct answer is probably to read `statusCode|lengthPrefix` from the input stream, then wrap it with `SnappyFramedInputStream` and use that to read until the end of that message (even if it's received across multiple Netty reads).  But there's still a risk the snappy stream read past the end of the message which would consume the status code and length prefix of the next message.  So I think we're always dependent on `SnappyFramedInputStream` not doing any read ahead regardless. If that becomes an issue we may need to contribute to the upstream project or fork it.

Reply:
> The completely correct answer is probably to read statusCode|lengthPrefix from the input stream, then wrap it with SnappyFramedInputStream and use that to read until the end of that message (even if it's received across multiple Netty reads).

That's pretty much what we're doing now.  But yeah, if snappy reads too much that's still a problem.  Though since we're asking it for a specific number of bytes which should correspond to the full uncompressed payload, it doesn't really have a reason to read beyond the current chunk unless the message is malformed.